? examples/jogottt
? examples/xadrez
? examples/dopecmd/l4/OBJ-x86_586-l4v2
? examples/dopelog/src/OBJ-x86_586-l4v2
? examples/minitop/l4/OBJ-x86_586-l4v2
? examples/presenter/l4/OBJ-x86_586-l4v2
? examples/run/l4/OBJ-x86_586-l4v2
? examples/textedit/l4/OBJ-x86_586-l4v2
? examples/vscrtest/data/OBJ-x86_586-l4v2
? examples/vscrtest/l4/OBJ-x86_586-l4v2
? examples/vtexttest/l4/OBJ-x86_586-l4v2
? examples/yuvtest/l4/OBJ-x86_586-l4v2
? idl/OBJ-x86-l4v2
? lib/dope/l4/OBJ-x86_586-l4v2
? lib/term/l4/OBJ-x86_586-l4v2
? lib/vscreen/l4/OBJ-x86_586-l4v2
? server/common/OBJ-x86_586-l4v2
? server/gfx/OBJ-x86_586-l4v2
? server/idl/OBJ-x86-l4v2
? server/l4/OBJ-x86_586-l4v2
Index: include/dopelib.h
===================================================================
RCS file: /home/remote-cvs/l4/pkg/dope/include/dopelib.h,v
retrieving revision 1.12
diff -u -r1.12 dopelib.h
--- include/dopelib.h	26 Aug 2004 14:44:28 -0000	1.12
+++ include/dopelib.h	11 Oct 2004 13:27:06 -0000
@@ -16,6 +16,10 @@
 #ifndef __DOPE_INCLUDE_DOPELIB_H_
 #define __DOPE_INCLUDE_DOPELIB_H_
 
+#ifdef __cplusplus
+extern "C" {
+#endif
+
 #define EVENT_TYPE_UNDEFINED    0
 #define EVENT_TYPE_COMMAND      1
 #define EVENT_TYPE_MOTION       2
@@ -153,6 +157,7 @@
  * \param app_id  DOpE application id
  */
 extern void dope_eventloop(long app_id);
+extern void dope_eventloop_poll(long app_id);
 
 
 /*** REQUEST KEY OR BUTTON STATE ***
@@ -172,4 +177,8 @@
  */
 extern char dope_get_ascii(long app_id, long keycode);
 
+#ifdef __cplusplus
+}
+#endif
+
 #endif /* __DOPE_INCLUDE_DOPELIB_H_ */
Index: lib/dope/common/main.c
===================================================================
RCS file: /home/remote-cvs/l4/pkg/dope/lib/dope/common/main.c,v
retrieving revision 1.14
diff -u -r1.14 main.c
--- lib/dope/common/main.c	28 Sep 2004 17:02:51 -0000	1.14
+++ lib/dope/common/main.c	11 Oct 2004 13:27:06 -0000
@@ -64,34 +64,44 @@
 
 
 /*** HANDLE EVENTLOOP OF A DOpE CLIENT ***/
-void dope_eventloop(long id) {
+void dope_eventloop_internal(long id, int block) {
 	dope_event *e;
 	char  *bindarg;
-	while (1) {
-		dopelib_wait_event(id, &e, &bindarg);
 
-		/* test if cookie is valid */
-		if ((bindarg[0] == '#') && (bindarg[1] == '!')) {
+	if((!block) && (!dopelib_check_event(id, NULL, NULL))) return;
+	dopelib_wait_event(id, &e, &bindarg);
+
+	/* test if cookie is valid */
+	if ((bindarg[0] == '#') && (bindarg[1] == '!')) {
+	
+		/* determine callback adress and callback argument */
+		u32 num1 = 0;
+		u32 num2 = 0;
+		char  *s = bindarg + 3;
 		
-			/* determine callback adress and callback argument */
-			u32 num1 = 0;
-			u32 num2 = 0;
-			char  *s = bindarg + 3;
-			
-			while (*s == ' ') s++;
-			while ((*s != 0) && (*s != ' ')) {
-				num1 = (num1 << 4) + (*s & 15);
-				if (*(s++) > '9') num1 += 9;
-			}
-			if (*(s++) == 0) return;
-			while ((*s != 0) && (*s != ' ')) {
-				num2 = (num2 << 4) + (*s&15);
-				if (*(s++) > '9') num2 += 9;
-			}
-			
-			/* call callback routine */
-			if (num1) ((void (*)(dope_event *, void *))num1)(e, (void *)num2);
+		while (*s == ' ') s++;
+		while ((*s != 0) && (*s != ' ')) {
+			num1 = (num1 << 4) + (*s & 15);
+			if (*(s++) > '9') num1 += 9;
+		}
+		if (*(s++) == 0) return;
+		while ((*s != 0) && (*s != ' ')) {
+			num2 = (num2 << 4) + (*s&15);
+			if (*(s++) > '9') num2 += 9;
 		}
+		
+		/* call callback routine */
+		if (num1) ((void (*)(dope_event *, void *))num1)(e, (void *)num2);
+	}
+}
+
+void dope_eventloop_poll(long id) {
+	dope_eventloop_internal(id, 0);
+}
+
+void dope_eventloop(long id) {
+	while(1) {
+		dope_eventloop_internal(id, 1);
 	}
 }
 
Index: lib/dope/dice/events.c
===================================================================
RCS file: /home/remote-cvs/l4/pkg/dope/lib/dope/dice/events.c,v
retrieving revision 1.10
diff -u -r1.10 events.c
--- lib/dope/dice/events.c	28 Sep 2004 17:02:51 -0000	1.10
+++ lib/dope/dice/events.c	11 Oct 2004 13:27:06 -0000
@@ -100,3 +100,11 @@
 	if (bindarg_out) *bindarg_out = &app->bindarg_queue[curr][0];
 	if (e_out) *e_out = &app->event_queue[curr];
 }
+
+int dopelib_check_event(int id, dope_event **e_out,char **bindarg_out) {
+	/* 0: no event, 1: got event */
+	struct dopelib_app *app;
+	app = dopelib_apps[id];
+	return !(app->first == app->last);
+}
+
Index: lib/dope/include/events.h
===================================================================
RCS file: /home/remote-cvs/l4/pkg/dope/lib/dope/include/events.h,v
retrieving revision 1.1
diff -u -r1.1 events.h
--- lib/dope/include/events.h	26 Mar 2003 23:34:55 -0000	1.1
+++ lib/dope/include/events.h	11 Oct 2004 13:27:07 -0000
@@ -1 +1,2 @@
 extern void dopelib_wait_event(int app_id, dope_event **e_out, char **bindarg_out);
+extern int dopelib_check_event(int app_id, dope_event **e_out, char **bindarg_out);
